File: android/src/main/java/com/jh/bluetooth/RNBleManager.kt
package com.jh.bluetooth

import android.content.Context
import com.desay.dsbluetooth.data.DSBLEAutoType
import com.desay.dsbluetooth.data.DSBLEBandFuncType
import com.desay.dsbluetooth.data.DSBLEDeviceType
import com.desay.dsbluetooth.data.model.DSBLESyncData
import com.desay.dsbluetooth.device.band.DSBLEBindCallback
import com.desay.dsbluetooth.device.band.DSBLESyncCallback
import com.desay.dsbluetooth.manager.keep.BLEAPIManager
import com.desay.dsbluetooth.manager.keep.DSBLECallback
import com.desay.dsbluetooth.manager.keep.DSBLEDevice
import com.facebook.react.bridge.*
import android.content.Intent
import com.desay.dsbluetooth.device.DSBLEDFUCallback
import com.facebook.react.bridge.ReactContext
import java.io.File


interface RNBleManagerDelegate {
    fun dispatchEvent(name: String, value: Any?)
}

/**
 * 蓝牙底层管理器
 */
class RNBleManager(enableLog: Boolean, private val context: Context) : DSBLECallback {

    private var bleManager: BLEAPIManager
    var delegate: RNBleManagerDelegate? = null
    private var syncHandler: RNBleManager.SyncHandler? = null
    private var bindHandler: RNBleManager.BindHandler? = null
    private var otaHandler: RNBleManager.OtaHandler? = null

    init {
        BLEAPIManager.context = context
        BLEAPIManager.otaClass = getMainActivityClass(context)!!::class.java
        if (enableLog) {
            BLEAPIManager.setLogOnWithCode("^DS*&0IJ#_jiKLPURIFIT$")
//            BLEAPIManager.logOut = { log ->
//                println(log)
//            }
        }
        bleManager = BLEAPIManager.instance
        bindHandler = this.BindHandler()
        syncHandler = this.SyncHandler()
        otaHandler = this.OtaHandler()
        bleManager.callback = this
    }

    private fun getMainActivityClass(context: Context): Class<*>? {
        val packageName = context.packageName
        val launchIntent = context.packageManager.getLaunchIntentForPackage(packageName)
        val className = launchIntent!!.component!!.className
        try {
            return Class.forName(className)
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
            return null
        }
    }

    // <editor-fold defaultstate="collapsed" desc="Module方法">
    fun destroy() {
        bindHandler = null
        syncHandler = null
        otaHandler = null
    }

    fun sdkVersion(promise: Promise) {
        val version = BLEAPIManager.version()
        if (version.isNotEmpty()) {
            promise.resolve(version)
        } else {
            promise.reject("0", "version is null", null)
        }
    }

    fun enableBle() {
        bleManager.enableBle()
    }

    fun disableBle() {
        bleManager.disableBle()
    }

    fun enableDebug(isDebug: Boolean) {
        BLEAPIManager.debug = isDebug
    }

    fun startScan(filter: Set<String>?) {
        bleManager.scan(null, filter)
    }

    fun stopScan() {
        bleManager.stopScan()
    }

    fun connect(mac: String?, autoConnect: Boolean) {
        bleManager.connectWithMac(mac ?: "", autoConnect)
    }

    fun disconnect(mac: String?) {
        bleManager.disconnectWithMac(mac ?: "")
    }

    fun state(promise: Promise) {
        promise.resolve(bleManager.isOn)
    }

    private fun bandNotify() {
        bleManager.band?.notify = { type, data ->
            this.handleNotify(type, data)
        }
    }

    fun bandSend(command: Int, data: Dynamic?, promise: Promise) {
        val sendType = RNBleSendType.values().firstOrNull { it.type == command }
        sendType?.let {
            val sendCommand = toNativeBandCommandType(it)
            val sendData = toNativeBandSendValue(it, data)
            bleManager.band?.makeFunc(sendCommand, sendData) { result, success ->
                if (success) {
                    val resultData = toJSBandValue(it, result)
                    promise.resolve(resultData)
                } else {
                    promise.reject("${BleErrorCode.UnknownError.code}", "Timeout", null)
                }
            }
        }
    }

    fun bandBind() {
        bleManager.band?.bindCallback = this.bindHandler
        bleManager.band?.makeFunc(DSBLEBandFuncType.bind)
    }

    fun bandSyncData() {
        bleManager.band?.syncCallback = this.syncHandler
        bleManager.band?.makeFunc(DSBLEBandFuncType.sync)
    }

    fun bandDfu(dfuFile: ReadableMap ,promise: Promise) {
        val dfuSet = DSBLEObjectBuilder.createDFUSet(dfuFile.toHashMap())
        bleManager.handleDFU(dfuSet,context)
        bleManager.dfuCallback = this.otaHandler
        promise.resolve(true)
    }
    // </editor-fold>

    // <editor-fold defaultstate="collapsed" desc="消息处理">
    private fun dispatchEvent(event: String, eventType: RNBleEventTypes, value: Any, error: RNBleError?) {
        val eventData = handleEvent(eventType, value, error)
        delegate?.dispatchEvent(event, eventData)
    }

    private fun handleEvent(eventType: RNBleEventTypes, value: Any, error: RNBleError?): WritableMap {
        val writableMap = Arguments.createMap()
        writableMap.putInt("event", eventType.type)
        if (error != null) {
            writableMap.putMap("error", error.asJSObject())
        } else {
            writableMap.putNull("error")
        }
        when {
            value as? WritableMap != null -> writableMap.putMap("data", value)
            value as? Int != null -> writableMap.putInt("data", value)
            value as? Double != null -> writableMap.putDouble("data", value)
            value as? String != null -> writableMap.putString("data", value)
            value as? Boolean != null -> writableMap.putBoolean("data", value)
            value as? WritableArray != null -> writableMap.putArray("data", value)
            else -> writableMap.putNull("data")
        }
        return writableMap
    }

    private fun handleSyncData(eventType: RNBleSyncTypes, value: Any?): WritableMap {
        val writableMap = Arguments.createMap()
        writableMap.putInt("type", eventType.type)
        when (eventType) {
            RNBleSyncTypes.Progress -> writableMap.putDouble("data", (value as? Double ?: 0.0))
            RNBleSyncTypes.Data -> {
                if (value as? DSBLESyncData != null) {
                    writableMap.putMap("data", value.asJSObject())
                } else {
                    writableMap.putNull("data")
                }
            }
            else -> writableMap.putNull("data")
        }
        return writableMap
    }

    private fun handleNotify(notifyType: DSBLEAutoType, data: Any?): WritableMap {
        val writableMap = Arguments.createMap()
        writableMap.putInt("type", notifyType.ordinal)
        when {
            data as? WritableMap != null -> writableMap.putMap("data", data)
            data as? Int != null -> writableMap.putInt("data", data)
            data as? Double != null -> writableMap.putDouble("data", data)
            data as? String != null -> writableMap.putString("data", data)
            data as? Boolean != null -> writableMap.putBoolean("data", data)
            data as? WritableArray != null -> writableMap.putArray("data", data)
            else -> writableMap.putNull("data")
        }
        return writableMap
    }

    private fun handleOta(otaType: RNOtaTypes, data: Any?) : WritableMap {
        val writableMap = Arguments.createMap()
        writableMap.putInt("type", otaType.type)
        when {
            data as? WritableMap != null -> writableMap.putMap("data", data)
            data as? Int != null -> writableMap.putInt("data", data)
            data as? Double != null -> writableMap.putDouble("data", data)
            data as? String != null -> writableMap.putString("data", data)
            data as? Boolean != null -> writableMap.putBoolean("data", data)
            data as? WritableArray != null -> writableMap.putArray("data", data)
            else -> writableMap.putNull("data")
        }
        return writableMap
    }

    // </editor-fold>

    // <editor-fold defaultstate="collapsed" desc="DSBLECallback接口实现">
    override fun onBLEStateChange(on: Boolean) {
        dispatchEvent(RNBleEvent.bleEvent, RNBleEventTypes.StateEvent, on, null)
    }

    override fun onConnectDFUDevice(device: DSBLEDevice) {
        dispatchEvent(RNBleEvent.bleEvent, RNBleEventTypes.OtaEvent, device.asJSObject(), null)
    }

    override fun onConnectDevice(device: DSBLEDevice) {
        if (device.type == DSBLEDeviceType.band) {
            this.bandNotify()
        }
        dispatchEvent(RNBleEvent.bleEvent, RNBleEventTypes.ConnectResultEvent, device.asJSObject(), null)
    }

    override fun onDisconnectDevice(device: DSBLEDevice) {
        dispatchEvent(RNBleEvent.bleEvent, RNBleEventTypes.DisconnectResultEvent, device.asJSObject(), null)
    }

    override fun onDiscoverDevice(device: DSBLEDevice) {
        dispatchEvent(RNBleEvent.scanEvent, RNBleEventTypes.ScanResultEvent, device.asJSObject(), null)
    }
    // </editor-fold>

    // <editor-fold defaultstate="collapsed" desc="实现DSBLEBindCallback"> 
    inner class BindHandler : DSBLEBindCallback {
        override fun onFailed() {
            dispatchEvent(RNBleEvent.bindEvent, RNBleEventTypes.BindEvent, 3, null)
        }

        override fun onStart() {
            dispatchEvent(RNBleEvent.bindEvent, RNBleEventTypes.BindEvent, 1, null)
        }

        override fun onSuccess() {
            dispatchEvent(RNBleEvent.bindEvent, RNBleEventTypes.BindEvent, 2, null)
        }
    }
    // </editor-fold> 

    // <editor-fold defaultstate="collapsed" desc="实现DSBLESyncCallback">
    inner class SyncHandler : DSBLESyncCallback {
        override fun onSuccess(syncData: DSBLESyncData?) {
            dispatchEvent(RNBleEvent.syncEvent, RNBleEventTypes.SyncEvent, handleSyncData(RNBleSyncTypes.Data, syncData), null)
        }

        override fun onWillStart() {
            dispatchEvent(RNBleEvent.syncEvent, RNBleEventTypes.SyncEvent, handleSyncData(RNBleSyncTypes.Start, null), null)
        }

        override fun onWillEnd() {
            dispatchEvent(RNBleEvent.syncEvent, RNBleEventTypes.SyncEvent, handleSyncData(RNBleSyncTypes.End, null), null)
        }

        override fun onSyncProgress(progress: Double) {
            dispatchEvent(RNBleEvent.syncEvent, RNBleEventTypes.SyncEvent, handleSyncData(RNBleSyncTypes.Progress, null), null)
        }

        override fun onFailed() {
            dispatchEvent(RNBleEvent.syncEvent, RNBleEventTypes.SyncEvent, handleSyncData(RNBleSyncTypes.Error, null), null)
        }
    }
    // </editor-fold>

    // <editor-fold defaultstate="collapsed" desc="实现DSBLEDFUCallback">
    inner class OtaHandler : DSBLEDFUCallback {
        override fun onOTAError(error: String) {
            dispatchEvent(RNBleEvent.otaEvent, RNBleEventTypes.OtaEvent, handleOta(RNOtaTypes.Error,error), null)
            bleManager.dfuCallback = null
        }

        override fun onOTAProgress(progress: Double) {
            dispatchEvent(RNBleEvent.otaEvent, RNBleEventTypes.OtaEvent, handleOta(RNOtaTypes.Progress,progress), null)
        }

        override fun onOTAStart() {
            dispatchEvent(RNBleEvent.otaEvent, RNBleEventTypes.OtaEvent, handleOta(RNOtaTypes.Start,null), null)
        }

        override fun onOTASuccess() {
            dispatchEvent(RNBleEvent.otaEvent, RNBleEventTypes.OtaEvent, handleOta(RNOtaTypes.End,null), null)
            bleManager.dfuCallback = null
        }
    }
    // </editor-fold>

}

File: ios/RNBleManager.swift
//
//  RNBleManager.swift
//  RNDsbluetooth
//
//  Created by Jaysonj on 2019/2/17.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation
import CoreBluetooth

@objc
public protocol RNBleManagerDelegate {
    func dispatchEvent(_ name: String, value: Any?)
}

@objc
class RNBleManager : NSObject {
    
    private let bleManager: BLEAPIManager
    
    @objc
    public var delegate: RNBleManagerDelegate?
    
    @objc
    public init(restoreStateIdentifier: String?, enableLog: Bool) {
        BLEAPIManager.restoreKey = restoreStateIdentifier
        bleManager = BLEAPIManager.shared
        super.init()
        if (enableLog) {
            bleManager.setLogOn(withCode: "^DS*&0IJ#_jiKLPURIFIT$")
            bleManager.logOut = { (log) in
                print("\(log)")
            }
        }
        bleManager.delegate = self
    }
    
    @objc
    public func sdkVersion(_ resole: @escaping Resolve, reject:  @escaping Reject) {
        let version = BLEAPIManager.shared.version()
        if version.count > 0 {
            resole(version)
        }else {
            reject("0","version is null",nil)
        }
    }
    
    @objc
    public func startScan(filter: Set<String>?, options: [String: Any]?) {
        // iOS handles allowDuplicates option to receive more scan records.
        var bleOptions = [String:Any]()
        if let options = options {
            if let allowDuplicates = options["allowDuplicates"] as? Bool {
                bleOptions[CBCentralManagerScanOptionAllowDuplicatesKey] = allowDuplicates
            }
        }
        bleManager.scan(withServices: nil, options: options != nil ? bleOptions : nil, filterNames: filter)
    }
    
    @objc
    public func stopScan() {
        bleManager.stopScan()
    }
    
    @objc
    public func enableDebug(_ debug: Bool) {
        bleManager.debug = debug;
    }
    
    @objc
    public func connect(identifier: String?, mac: String?) {
        if let connectID = identifier {
            bleManager.connect(withIdentifier: connectID)
        }else {
            bleManager.connect(withMac: mac)
        }
    }
    
    @objc
    public func disconnect(identifier: String?) {
        bleManager.disconnect(withIdentifier: identifier)
    }
    
    @objc
    public func state(_ resole: @escaping Resolve){
        return resole(bleManager.bleState == .on)
    }
    
    @objc
    public func bandSend(_ command: Int, data: Any?, resole: @escaping Resolve, reject:  @escaping Reject) {
        if let sendTypes = RNBleSendType.init(rawValue: command) {
            // 转化为SDK对象的指令和数据格式
            let bandCommand = sendTypes.toNativeBandCommandType()
            let bandData = sendTypes.toNativeBandSendValue(data)
            bleManager.band?.makeFunc(bandCommand, data: bandData, callback: {(result, success, error) in
                if success {
                    // 转化SDK结果为JS对象的数据格式
                    let returnValue = sendTypes.toJSBandValue(result)
                    resole(returnValue)
                }else {
                    if let err = error {
                        reject("\(BleErrorCode.OperationTimedOut.rawValue)",err.localizedDescription,err as NSError)
                    }else {
                        reject("\(BleErrorCode.UnknownError.rawValue)","Timeout",RNBleError.init(errorCode: BleErrorCode.UnknownError, reason: "Timeout", attErrorCode: nil, iosErrorCode: nil) as NSError)
                    }
                }
            })
        }
    }
    
    @objc
    public func bandSyncData() {
        bleManager.band?.syncDelegate = self
        bleManager.band?.makeFunc(.sync, data: nil, callback: nil)
    }
    
    @objc
    public func bandBind() {
        bleManager.band?.bindDelegate = self
        bleManager.band?.makeFunc(.bind, data: nil, callback: nil)
    }
    
    func bandNotify() {
        bleManager.band?.notify = { [weak self] (type, data, error) in
            let result = self?.handlerNotify(autoType: type, data: data)
            self?.dispatchEvent(RNBleEvent.notifyEvent, eventType: .notifyEvent, value: result!, error: nil)
        }
    }
    
    @objc
    public func destroy() {
        bleManager.band?.disconnect()
        bleManager.delegate = nil
    }
    
    @objc
    public func bandDfu(_ dfuFile: Any, resole: @escaping Resolve, reject:  @escaping Reject) {
        if let dfuData = RNBleSendType.dfu.toNativeBandSendValue(dfuFile) {
            if let dfu = dfuData as? DSBLEDFUSet {
                bleManager.dfuDelegate = self
                bleManager.handleDFU(With: dfu)
                resole(true)
            }else {
                reject("\(BleErrorCode.BluetoothUnsupported.rawValue)","无法识别的DFU文件格式",RNBleError.init(errorCode: BleErrorCode.BluetoothUnsupported, reason: "无法识别的DFU文件格式", attErrorCode: nil, iosErrorCode: nil) as NSError)
            }
        }else {
            reject("\(BleErrorCode.BluetoothUnsupported.rawValue)","无法识别的DFU文件格式",RNBleError.init(errorCode: BleErrorCode.BluetoothUnsupported, reason: "无法识别的DFU文件格式", attErrorCode: nil, iosErrorCode: nil) as NSError)
        }
    }
}

// MARK: - 处理消息发送
extension RNBleManager {
    
    fileprivate func dispatchEvent(_ event: String, eventType: RNBleEventTypes, value: Any, error: RNBleError?) {
        let eventValue = handlerEvent(event, eventType: eventType, value: value, error: error)
        delegate?.dispatchEvent(event, value: eventValue)
    }
    
    fileprivate func handlerNotify(autoType: DSBLEAutoType,data: Any?) -> [String : Any] {
        return [
            "type": autoType.rawValue,
            "data": data ?? NSNull(),
        ]
    }
    
    /// 封装发送事件格式
    ///
    /// - Parameters:
    ///   - event: 事件类型
    ///   - value: 事件数据
    fileprivate func handlerEvent(_ event: String,eventType: RNBleEventTypes, value: Any, error: RNBleError?) -> [String : Any] {
        return [
            "event": eventType.rawValue,
            "error": error?.asJSObject() ?? NSNull(),
            "data": value,
        ]
    }
}

// MARK: - DSBLEScanConnectDelegate
extension RNBleManager: DSBLEScanConnectDelegate {
    
    func notifyState(_ state: BLEManagerState) {
        /// 系统蓝牙变更
        dispatchEvent(RNBleEvent.bleEvent, eventType: .stateEvent, value: state == .on, error: nil)
    }
    
    func didDiscoverDevice(_ device: DSBLEDevice) {
        /// 扫描获取到的设备
        print("didDiscoverDevice")
        device.debug()
        dispatchEvent(RNBleEvent.scanEvent,eventType: .scanResultEvent, value: device.asJSObject(), error: nil)
    }
    
    func didConnectDevice(_ device: DSBLEDevice) {
        /// 连接到的设备
        print("didConnectDevice")
        device.debug()
        dispatchEvent(RNBleEvent.bleEvent,eventType: .connectResultEvent, value: device.asJSObject(), error: nil)
        if device.type == .band {
            bandNotify()
        }
    }
    
    func didDisconnectDevice(_ device: DSBLEDevice) {
        /// 断开连接的设备
        print("didDisconnectDevice")
        device.debug()
        dispatchEvent(RNBleEvent.bleEvent,eventType: .disconnectResultEvent, value: device.asJSObject(), error: nil)
    }
    
    func connectDFUDevice(_ device: DSBLEDevice) {
        /// OTA 设备
        print("connectDFUDevice")
        device.debug()
        dispatchEvent(RNBleEvent.bleEvent,eventType: .otaEvent, value: device.asJSObject(), error: nil)
    }
    
    func restoreStateDict(_ dict: [String : Any]) {
        /// 恢复设备
        //        dispatchEvent(RNBleEvent.restoreStateEvent, value: dict, error: nil)
    }
}


// MARK: - 绑定事件
extension RNBleManager : DSBLEBindDelegate {
    
    func handleBindState(_ state: DSBLEBindingState) {
        dispatchEvent(RNBleEvent.bindEvent, eventType: .bindEvent, value: state.rawValue, error: nil)
    }
}

extension RNBleManager : DSBLESyncDelegate {
    
    func syncProgress(progress: Float) {
        
        dispatchEvent(RNBleEvent.syncEvent, eventType: .syncEvent, value: handlerSyncData(eventType: .progress, value: progress), error: nil)
    }
    
    func syncWillStart() {
        dispatchEvent(RNBleEvent.syncEvent, eventType: .syncEvent, value: handlerSyncData(eventType: .start, value: nil), error: nil)
    }
    
    func syncWillEnd() {
        dispatchEvent(RNBleEvent.syncEvent, eventType: .syncEvent, value: handlerSyncData(eventType: .end, value: nil), error: nil)
    }
    
    func syncError() {
        dispatchEvent(RNBleEvent.syncEvent, eventType: .syncEvent, value: handlerSyncData(eventType: .error, value: nil), error: nil)
    }
    
    func syncFinished(syncResultData syncData: DSBLESyncData?) {
        dispatchEvent(RNBleEvent.syncEvent, eventType: .syncEvent, value: handlerSyncData(eventType: .data, value: syncData), error: nil)
    }
    
    private func handlerSyncData(eventType: RNBleSyncTypes, value: Any?) -> [String : Any] {
        var data: Any? = nil
        switch eventType {
        case .progress:
            data = value ?? 0.0
        case .data:
            data = (value as? DSBLESyncData)?.asJSObject() ?? [:]
        default:
            data = NSNull()
        }
        return [
            "type": eventType.rawValue,
            "data": data!
        ]
    }
}

extension RNBleManager: DSBLEDFUDelegate {
    
    func DFUProgress(_ progress: Float) {
        dispatchEvent(RNBleEvent.otaEvent, eventType: .otaEvent, value: handlerOtaData(eventType: .progress, value: progress), error: nil)
    }
    
    func DFUStart() {
        dispatchEvent(RNBleEvent.otaEvent, eventType: .otaEvent, value: handlerOtaData(eventType: .start, value: nil), error: nil)
    }
    
    func DFUEnd() {
        dispatchEvent(RNBleEvent.otaEvent, eventType: .otaEvent, value: handlerOtaData(eventType: .end, value: nil), error: nil)
        bleManager.dfuDelegate = nil
    }
    
    func DFUError(_ error: String) {
        dispatchEvent(RNBleEvent.otaEvent, eventType: .otaEvent, value: handlerOtaData(eventType: .error, value: error), error: nil)
        bleManager.dfuDelegate = nil
    }
    
    private func handlerOtaData(eventType: RNBleOtaTypes, value: Any?) -> [String : Any] {
        var data: Any? = nil
        switch eventType {
        case .progress:
            data = value ?? 0.0
        case .error:
            data = value ?? ""
        default:
            data = NSNull()
        }
        return [
            "type": eventType.rawValue,
            "data": data!
        ]
    }
}

